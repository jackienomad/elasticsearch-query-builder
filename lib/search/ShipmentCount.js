import { formatToTimeZone, parseFromTimeZone } from 'date-fns-timezone'

export function generateQuery(
  start,
  end,
  companyId,
  traderType,
  arrivalDateField,
  consigneeField,
  shipperField
) {
  // build query
  const query = {
    size: 0,
    query: {
      bool: {
        filter: [
          {
            bool: {
              must: []
            }
          },
          {
            range: {
              [arrivalDateField]: {
                gte: start
              }
            }
          },
          {
            range: {
              [arrivalDateField]: {
                lte: end
              }
            }
          }
        ]
      }
    }
  }

  // add trader type filter
  if (traderType === 'consignee' || traderType === 'internal') {
    query.query.bool.filter[0].bool.must.push({
      term: {
        [consigneeField]: companyId
      }
    })
  }

  if (traderType === 'shipper' || traderType === 'internal') {
    query.query.bool.filter[0].bool.must.push({
      term: {
        [shipperField]: companyId
      }
    })
  }

  return query
}

export class ShipmentCount {
  constructor() {}

  static buildQuery(
    fieldMap,
    country,
    shipmentType,
    companyId,
    arrivalDates,
    traderTypes
  ) {
    const body = []

    const arrivalDateField = fieldMap['ArrivalDate']['field']
    const consigneeField = fieldMap['ConsigneeId']['field']
    const shipperField = fieldMap['ShipperId']['field']

    const queries = {
      consignee: [],
      shipper: [],
      internal: []
    }

    arrivalDates.forEach(({ start, end }) => {
      traderTypes.forEach(traderType => {
        queries[traderType].push({ index: `${country}_${shipmentType}` })
        queries[traderType].push(
          generateQuery(
            start,
            end,
            companyId,
            traderType,
            arrivalDateField,
            consigneeField,
            shipperField
          )
        )
      })
    })

    body.push(...queries['consignee'])
    body.push(...queries['shipper'])
    body.push(...queries['internal'])

    return { body }
  }

  static formatResponse(esResponse, arrivalDates, traderTypes, timezone) {
    const traderTypeCount = traderTypes.length
    const esResponseCount = esResponse.responses.length
    const margin = esResponseCount / traderTypeCount
    const arrivalDatesCount = arrivalDates.length

    const formattedResponse = esResponse.responses.map((response, index) => ({
      from: +formatToTimeZone(
        parseFromTimeZone(
          arrivalDates[index % arrivalDatesCount].start * 1000,
          {
            timeZone: 'Africa/Accra'
          }
        ),
        'X',
        { timeZone: timezone }
      ),
      to: +formatToTimeZone(
        parseFromTimeZone(arrivalDates[index % arrivalDatesCount].end * 1000, {
          timeZone: 'Africa/Accra'
        }),
        'X',
        { timeZone: timezone }
      ),
      totalShipments: response.hits.total,
      traderType: traderTypes[Math.floor(index / margin)]
    }))
    return formattedResponse
  }
}

export default ShipmentCount
