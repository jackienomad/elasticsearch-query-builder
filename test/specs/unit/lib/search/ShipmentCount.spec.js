import ShipmentCount from 'qb/@/search/ShipmentCount'
import fieldMap from './dataset/sampleFieldMapping'
import { esResponse } from './dataset/shipmentCountElasticSearchResponse'

describe('class ShipmentCount', () => {
  describe('constructor', () => {
    it('should be constructible', () => {
      expect(() => new ShipmentCount()).not.toThrowError()
    })
  })

  describe('static method buildQuery', () => {
    it('should return a correct elasticsearch query', async () => {
      const country = 'us'
      const shipmentType = 'import'
      const companyId = '28a79d28abfe3f78fed5f541e9e28a6d'
      const arrivalDates = [
        {
          start: 1480762061,
          end: 1536192000
        }
      ]
      const traderTypes = ['consignee', 'shipper', 'internal']

      const query = ShipmentCount.buildQuery(
        fieldMap,
        country,
        shipmentType,
        companyId,
        arrivalDates,
        traderTypes
      )

      expect(query).toMatchSnapshot()
    })
  })

  describe('static method formatResponse', () => {
    it('should format the elasticsearch response to expected format', async () => {
      const formattedResponse = ShipmentCount.formatResponse(
        esResponse,
        [
          {
            start: 1480762061,
            end: 1536192000
          }
        ],
        ['consignee', 'shipper', 'internal'],
        'Asia/Manila'
      )
      expect(formattedResponse).toMatchSnapshot()
    })
  })
})
