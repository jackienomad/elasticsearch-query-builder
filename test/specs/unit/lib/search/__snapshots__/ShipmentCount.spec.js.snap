// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class ShipmentCount static method buildQuery should return a correct elasticsearch query 1`] = `
Object {
  "body": Array [
    Object {
      "index": "us_import",
    },
    Object {
      "query": Object {
        "bool": Object {
          "filter": Array [
            Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "Consignee._id": "28a79d28abfe3f78fed5f541e9e28a6d",
                    },
                  },
                ],
              },
            },
            Object {
              "range": Object {
                "Arrival Date": Object {
                  "gte": 1480762061,
                },
              },
            },
            Object {
              "range": Object {
                "Arrival Date": Object {
                  "lte": 1536192000,
                },
              },
            },
          ],
        },
      },
      "size": 0,
    },
    Object {
      "index": "us_import",
    },
    Object {
      "query": Object {
        "bool": Object {
          "filter": Array [
            Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "Shipper._id": "28a79d28abfe3f78fed5f541e9e28a6d",
                    },
                  },
                ],
              },
            },
            Object {
              "range": Object {
                "Arrival Date": Object {
                  "gte": 1480762061,
                },
              },
            },
            Object {
              "range": Object {
                "Arrival Date": Object {
                  "lte": 1536192000,
                },
              },
            },
          ],
        },
      },
      "size": 0,
    },
    Object {
      "index": "us_import",
    },
    Object {
      "query": Object {
        "bool": Object {
          "filter": Array [
            Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "Consignee._id": "28a79d28abfe3f78fed5f541e9e28a6d",
                    },
                  },
                  Object {
                    "term": Object {
                      "Shipper._id": "28a79d28abfe3f78fed5f541e9e28a6d",
                    },
                  },
                ],
              },
            },
            Object {
              "range": Object {
                "Arrival Date": Object {
                  "gte": 1480762061,
                },
              },
            },
            Object {
              "range": Object {
                "Arrival Date": Object {
                  "lte": 1536192000,
                },
              },
            },
          ],
        },
      },
      "size": 0,
    },
  ],
}
`;

exports[`class ShipmentCount static method formatResponse should format the elasticsearch response to expected format 1`] = `
Array [
  Object {
    "from": 1480790861,
    "to": 1536220800,
    "totalShipments": 27,
    "traderType": "consignee",
  },
  Object {
    "from": 1480790861,
    "to": 1536220800,
    "totalShipments": 0,
    "traderType": "shipper",
  },
  Object {
    "from": 1480790861,
    "to": 1536220800,
    "totalShipments": 0,
    "traderType": "internal",
  },
]
`;
